src/subscript.c

```diff
@@ -165,17 +165,21 @@ static SEXP dbl_cast_subscript(SEXP subscript,
       continue;
     }
 
-    int out_elt = (int) elt;
-
-    // Detect out-of-bounds and fractional numbers
-    if (elt > INT_MAX || out_elt != elt) {
+    if (!isfinite(elt) || elt > INT_MAX) {
```

Comment:
So we were already capturing `NA_real_` and `NaN` above with `isnan()`, and the existing `elt > INT_MAX` check should have been capturing any large double like `Inf` or `INT_MAX+1`, so I imagine the problem came from some kind of double value that was too small?

I think this check would still improperly allow a double of `INT_MIN - 1` through, which would still fail?

I think I would use `if (elt > INT_MAX || elt < INT_MIN + 1)`

This follows what I think base R does when converting a double to integer (see link below). `elt < INT_MIN + 1` should capture any doubles that are too small to fit in an integer, and also doesn't allow a double that is identical to `INT_MIN`, because that is equivalent to `NA_integer_`.

https://github.com/wch/r-source/blob/5a156a0865362bb8381dcd69ac335f5174a4f60c/src/main/cum.c#L47-L51

Reply:
Good point

Reply:
The existing `>` check occurred after conversion by the way, which would cause the Inf cast UB. So we still need to check for finiteness.